cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -663,16 +663,25 @@ public int getWorkerInstances(ProgramId id) {
   }
 
   @Override
-  public void removeApplication(ApplicationId id) {
-    LOG.trace("Removing application: namespace: {}, application: {}", id.getNamespace(),
-        id.getApplication());
+  public void removeApplication(ApplicationReference appRef) {
+    LOG.trace("Removing application: namespace: {}, application: {}", appRef.getNamespace(),
+        appRef.getApplication());
 
     TransactionRunners.run(transactionRunner, context -> {
-      getAppStateTable(context).deleteAll(id.getNamespaceId(), id.getApplication());
       AppMetadataStore metaStore = getAppMetadataStore(context);
-      metaStore.deleteApplication(id.getNamespace(), id.getApplication(), id.getVersion());
-      metaStore.deleteApplicationEditRecord(id.getAppReference());
-      metaStore.deleteProgramHistory(id.getNamespace(), id.getApplication(), id.getVersion());
+      metaStore.deleteApplication(appRef);
+      metaStore.deleteProgramHistory(appRef);
+    });
+  }
+
+  @Override
+  public void removeApplicationData(ApplicationReference appRef) {
+    LOG.trace("Removing application: namespace: {}, application: {}", appRef.getNamespace(),
+              appRef.getApplication());
+
+    TransactionRunners.run(transactionRunner, context -> {
+      getAppStateTable(context).deleteAll(appRef.getNamespaceId(), appRef.getApplication());
+      getAppMetadataStore(context).deleteApplicationEditRecord(appRef);
```

Comment:
I think this should be part of the metaStore.deleteApplication() method, it is not valid to have this row exist while there are no rows in the app spec table.